accounts/abi/bind/topics.go

```diff
@@ -103,151 +103,61 @@ func makeTopics(query ...[]interface{}) ([][]common.Hash, error) {
-		if !arg.Indexed {
-			return errors.New("non-indexed field in topic reconstruction")
-		}
+	store := func(arg abi.Argument, reconstr interface{}) {
 		field := reflect.ValueOf(out).Elem().FieldByName(capitalise(arg.Name))
-
-		// Try to parse the topic back into the fields based on primitive types
-		switch field.Kind() {
-		case reflect.Bool:
-			if topics[0][common.HashLength-1] == 1 {
-				field.Set(reflect.ValueOf(true))
-			}
-		case reflect.Int8:
-			num := new(big.Int).SetBytes(topics[0][:])
-			field.Set(reflect.ValueOf(int8(num.Int64())))
-
-		case reflect.Int16:
-			num := new(big.Int).SetBytes(topics[0][:])
-			field.Set(reflect.ValueOf(int16(num.Int64())))
-
-		case reflect.Int32:
-			num := new(big.Int).SetBytes(topics[0][:])
-			field.Set(reflect.ValueOf(int32(num.Int64())))
-
-		case reflect.Int64:
-			num := new(big.Int).SetBytes(topics[0][:])
-			field.Set(reflect.ValueOf(num.Int64()))
-
-		case reflect.Uint8:
-			num := new(big.Int).SetBytes(topics[0][:])
-			field.Set(reflect.ValueOf(uint8(num.Uint64())))
-
-		case reflect.Uint16:
-			num := new(big.Int).SetBytes(topics[0][:])
-			field.Set(reflect.ValueOf(uint16(num.Uint64())))
-
-		case reflect.Uint32:
-			num := new(big.Int).SetBytes(topics[0][:])
-			field.Set(reflect.ValueOf(uint32(num.Uint64())))
-
-		case reflect.Uint64:
-			num := new(big.Int).SetBytes(topics[0][:])
-			field.Set(reflect.ValueOf(num.Uint64()))
-
-		default:
-			// Ran out of plain primitive types, try custom types
-
-			switch field.Type() {
-			case reflectHash: // Also covers all dynamic types
-				field.Set(reflect.ValueOf(topics[0]))
-
-			case reflectAddress:
-				var addr common.Address
-				copy(addr[:], topics[0][common.HashLength-common.AddressLength:])
-				field.Set(reflect.ValueOf(addr))
-
-			case reflectBigInt:
-				num := new(big.Int).SetBytes(topics[0][:])
-				if arg.Type.T == abi.IntTy {
-					if num.Cmp(abi.MaxInt256) > 0 {
-						num.Add(abi.MaxUint256, big.NewInt(0).Neg(num))
-						num.Add(num, big.NewInt(1))
-						num.Neg(num)
-					}
-				}
-				field.Set(reflect.ValueOf(num))
-
-			default:
-				// Ran out of custom types, try the crazies
-				switch {
-				// static byte array
-				case arg.Type.T == abi.FixedBytesTy:
-					reflect.Copy(field, reflect.ValueOf(topics[0][:arg.Type.Size]))
-				default:
-					return fmt.Errorf("unsupported indexed type: %v", arg.Type)
-				}
-			}
-		}
-		topics = topics[1:]
+		field.Set(reflect.ValueOf(reconstr))
 	}
-	return nil
+	return parseTopicIntoFunc(fields, topics, store)
 }
 
 // parseTopicsIntoMap converts the indexed topic field-value pairs into map key-value pairs
 func parseTopicsIntoMap(out map[string]interface{}, fields abi.Arguments, topics []common.Hash) error {
+	store := func(arg abi.Argument, reconstr interface{}) {
+		out[arg.Name] = reconstr
+	}
+	return parseTopicIntoFunc(fields, topics, store)
+}
+
+// parseTopicIntoFunc converts the indexed topic field-value pairs and stores them using the
+// provided store function.
```

Comment:
Actually, instead of Store/Func in names, I would use the word "setter", which makes the code much easier to understand.

Reply:
@gballet `parseTopicsWithSetter` or `parseTopicsIntoSetter`?